cmd/panfs-multipart.go

```diff
@@ -73,31 +100,42 @@ func (fs *PANFSObjects) decodePartFile(name string) (partNumber int, etag string
 }
 
 // Appends parts to an appendFile sequentially.
-func (fs *PANFSObjects) backgroundAppend(ctx context.Context, bucket, object, uploadID string) {
-	fs.appendFileMapMu.Lock()
+func (fs *PANFSObjects) backgroundAppend(ctx context.Context, bucket, object, uploadID string, need_flock bool) {
 	logger.GetReqInfo(ctx).AppendTags("uploadID", uploadID)
 	bucketPath, err := fs.getBucketPanFSPath(ctx, bucket)
 	if err != nil {
 		logger.LogIf(ctx, err, logger.Application)
 		return
 	}
-
+	fs.appendFileMapMu.Lock()
 	file := fs.appendFileMap[uploadID]
 	if file == nil {
-		file = &panfsAppendFile{
-			filePath: pathJoin(bucketPath, panfsS3TmpDir, fs.nodeDataSerial, bgAppendsDirName, fmt.Sprintf("%s.%s", uploadID, mustGetUUID())),
+		file = fs.getPanFSMultipartAppendFile(bucketPath, object, uploadID)
+		if file == nil {
+			fs.appendFileMapMu.Unlock()
+			return
 		}
 		fs.appendFileMap[uploadID] = file
 	}
 	fs.appendFileMapMu.Unlock()
 
-	file.Lock()
-	defer file.Unlock()
+	// No need to take a lock when bgAppend called from CompleteMultipartUpload which has been already taken the lock
+	if need_flock {
+		if !file.flock.TryLock() {
```

Comment:
I would add comment that there are two main scenarios when we cannot take this lock:

1. Multipart upload was completed or aborted, lock file was removed. So enqueued goroutine can just exit and do nothing. File append map structure that we added above will be cleaned by timeout.
2. Another server/thread is doing background append now, so we are not waiting for it to finish.